feat: adds sort widget to search manager and library component page [FC-0059]

pomegranited opened the pull request:
Please use this PR for code review.
See https://github.com/openedx/frontend-app-course-authoring/pull/1147 for testing instructions.

Private-ref: [FAL-3758](https://tasks.opencraft.com/browse/FAL-3758)

yusuf-musleh committed:
feat: Add lib v2/legacy tabs in studio home

When lib mode is set to "mixed", both "Libraries" and "Legacy Libraries" tabs are show in the Studio Home. When "Libraries" is clicked, v2 libraries are fetched, when "Legacy Libraries" is clicked, v1 libraries are fetched.

When lib mode is set to "v1 only" or "v2 only", only one tab "Libraries" is show and only the respective libraries are fetched when the tab is clicked.

yusuf-musleh committed:
feat: Add lib v2/legacy tabs in studio home

When lib mode is set to "mixed", both "Libraries" and "Legacy Libraries" tabs are show in the Studio Home. When "Libraries" is clicked, v2 libraries are fetched, when "Legacy Libraries" is clicked, v1 libraries are fetched.

When lib mode is set to "v1 only" or "v2 only", only one tab "Libraries" is show and only the respective libraries are fetched when the tab is clicked.

yusuf-musleh committed:
feat: Add `LIBRARY_MODE` config variable

This is to switch between different library modes.

yusuf-musleh committed:
feat: Add `LIBRARY_MODE` config variable

This is to switch between different library modes.

yusuf-musleh committed:
feat: Add url paths/navigation for each tab

The path updates when selecting tabs, when accessing the url with the path directly it will open its respective tab. Navigating using the browser back/forward buttons is also supported.

yusuf-musleh committed:
feat: Add url paths/navigation for each tab

The path updates when selecting tabs, when accessing the url with the path directly it will open its respective tab. Navigating using the browser back/forward buttons is also supported.

yusuf-musleh committed:
feat: LibraryV2 redirect to lib mfe or placeholder

yusuf-musleh committed:
feat: LibraryV2 redirect to lib mfe or placeholder

yusuf-musleh committed:
feat: Add pagination support for lib v2s

yusuf-musleh committed:
feat: Add pagination support for lib v2s

yusuf-musleh committed:
fix: Redirect to placeholder create lib in v2/mixed disabled mfe

yusuf-musleh committed:
fix: Redirect to placeholder create lib in v2/mixed disabled mfe

yusuf-musleh committed:
temp: This removes TS code to get tests to run

This commit is temporary as the current frontend build system in tests doesnt support TS syntax. That should be fixed soon, and this commit should be removed.

yusuf-musleh committed:
temp: This removes TS code to get tests to run

This commit is temporary as the current frontend build system in tests doesnt support TS syntax. That should be fixed soon, and this commit should be removed.

yusuf-musleh committed:
test: Update existing tests to support changes

yusuf-musleh committed:
test: Update existing tests to support changes

yusuf-musleh committed:
temp: Rename .tsx -> .jsx & .ts -> .js for tests

This is a temporary commit since there are currently no webpack loaders that support tsx files in the test running. This commit should be removed once that is fixed upstream.

yusuf-musleh committed:
temp: Rename .tsx -> .jsx & .ts -> .js for tests

This is a temporary commit since there are currently no webpack loaders that support tsx files in the test running. This commit should be removed once that is fixed upstream.

yusuf-musleh committed:
fix: Fix lint issues

yusuf-musleh committed:
fix: Fix lint issues

rpenido committed:
feat: library home page bare bones

yusuf-musleh committed:
test: Add tests for new functionality

yusuf-musleh committed:
test: Add tests for new functionality

rpenido committed:
fix: update search modal for new library urls

yusuf-musleh committed:
refactor: Change /legacy-libraries -> /libraries-v1

rpenido committed:
fix: add i18n messages

rpenido committed:
fix: libraryAuthoring enabled check

rpenido committed:
Merge branch 'yusuf-musleh/lib-v2-tab-studio-home' into rpenido/fal-3753-library-home-page-bare-bones

rpenido committed:
fix: add Create Library placeholder

yusuf-musleh committed:
refactor: Remove hardcoded mfe path

rpenido committed:
refactor: rename .ts files to .js

yusuf-musleh committed:
feat: Add function to construct Lib Auth MFE URL

yusuf-musleh committed:
feat: Make URL /library-v1 when referencing legacy

yusuf-musleh committed:
fix: Add missing part of path

rpenido committed:
fix: type and lint errors

yusuf-musleh committed:
fix: Issue with destinationUrl

yusuf-musleh committed:
test: Add checks for Tab.eventKey in tests

yusuf-musleh committed:
fix: Revert card item url changes to keep simple

rpenido committed:
fix: add tests

rpenido committed:
Merge branch 'yusuf-musleh/lib-v2-tab-studio-home' into rpenido/fal-3753-library-home-page-bare-bones

rpenido committed:
fix: removing unused file

rpenido committed:
fix: add ts-check

rpenido committed:
fix: removing deleted file references

rpenido committed:
chore: trigger CI

ChrisChV committed:
feat: Initial structure of ComponentCard

rpenido committed:
fix: fixes from review

rpenido committed:
fix: fix default parameter syntax

ChrisChV committed:
feat: Connect TagCount, Colors and Icons with library components card

ChrisChV committed:
feat: Infinite scroll implemented on Components tab

rpenido committed:
fix: new library redirect

rpenido committed:
Revert "fix: new library redirect"

This reverts commit c63bc2fbf586a791770b3b63e3262c5c6bd542ec.

rpenido committed:
chore: trigger CI

ChrisChV committed:
style: Style nits on card component

rpenido committed:
fix: new library redirect

rpenido committed:
fix: remove warning on library home

rpenido committed:
fix: remove unused import

rpenido committed:
Merge branch 'master' into rpenido/fal-3753-library-home-page-bare-bones

rpenido committed:
fix: merging nits

rpenido committed:
fix: change version to slug in header

rpenido committed:
fix: header outline link for libraries

ChrisChV committed:
style: Add fetchNextPage to infinite scroll useEffect

ChrisChV committed:
feat: Variant added to LibraryComponents to show full components or a preview

ChrisChV committed:
style: Nits on lints and code

ChrisChV committed:
refactor: Create block-type-utils with block type constants

ChrisChV committed:
fix: Issues on course-unit tests

ChrisChV committed:
refactor: Calculate tagCount in LibraryComponents instead on serach index

rpenido committed:
Merge branch 'master' into rpenido/fal-3753-library-home-page-bare-bones

rpenido committed:
refactor: test renaming to ts

rpenido committed:
refactor: renaming files js -> ts

rpenido committed:
fix: update typescript code

rpenido committed:
Merge branch 'master' into rpenido/fal-3753-library-home-page-bare-bones

ChrisChV committed:
test: Added for LibraryComponents

rpenido committed:
fix: remove @ts-check

rpenido committed:
fix: remove unused file

rpenido committed:
feat: use search manager

rpenido committed:
refactor: cleanup code

rpenido committed:
fix: more cleanup

ChrisChV committed:
chore: Fix merge conflicts

yusuf-musleh committed:
feat: Adjust library home filter styles

rpenido committed:
Merge branch 'master' into rpenido/fal-3764-library-home-filter-by-content-type

rpenido committed:
fix: remove LibraryV2Placeholder

rpenido committed:
Merge branch 'master' into rpenido/fal-3753-library-home-page-bare-bones

rpenido committed:
fix: optional parameters

rpenido committed:
chore: trigger CI

ChrisChV committed:
refactor: Migrate block-type-utils to TS

ChrisChV committed:
chore: Fix merge conflicts

ChrisChV committed:
test: Fixt LibraryAuthoringPage test

rpenido committed:
Merge branch 'chris/FAL-3757-components-tab' into rpenido/fal-3764-library-home-filter-by-content-type

rpenido committed:
feat: fixing merge issues and tests

rpenido committed:
refactor: create search-manager feature

rpenido committed:
test: fix search-modal locator

rpenido committed:
Merge branch 'yusuf-musleh/library-home-filters-styling' into rpenido/fal-3764-library-home-filter-by-content-type

rpenido committed:
fix: fix scss files location

rpenido committed:
refactor: moving Stats

rpenido committed:
fix: remove Stats from library home

rpenido committed:
fix: remove loading state if already data is already loaded to avoid flickering

rpenido committed:
feat: library home page bare bones

yusuf-musleh committed:
feat: Add clear filter btn to filter widget

rpenido committed:
refactor: migrate Header to ts and fix some types

rpenido committed:
fix: add intl to tab names

rpenido committed:
fix: remove bottom border to align label with checkbox

rpenido committed:
fix: add intl to tab names

rpenido committed:
Merge branch 'rpenido/fal-3753-library-home-page-bare-bones' into rpenido/fal-3764-library-home-filter-by-content-type

rpenido committed:
Merge branch 'master' into rpenido/fal-3764-library-home-filter-by-content-type

rpenido committed:
fix: removing unused file

rpenido committed:
fix: removing unused file

pomegranited committed:
feat: adds sort widget to search manager and library component page

pomegranited committed:
fix: type error

Type 'ReactElement<any, string | JSXElementConstructor<any>>' is not assignable to type 'ComponentType<{}> | undefined'.

pomegranited commented:
@yusuf-musleh I will address this point on the ticket today by modifying the SearchManager:

> The sort option should be reflected in the URL, e.g. ?sort=created but should not create a new history entry, so pressing BACK in the browser goes back to the previous page, not the previous sort option.

yusuf-musleh commented:
@pomegranited This is looking great so far! I've tested it out and its working quite well. I like the use of an enum for the sort options, very clean. And this PR would definitely make things easier for me working on the "Recently Modified" section.

rpenido committed:
fix: move block type intl

pomegranited committed:
feat: store selected sort option in query string

pomegranited committed:
fix: better way to store sort params in query string

using a re-usable custom hook.

Also fixed type/lint errors/console warnings.

pomegranited committed:
Merge branch 'rpenido/fal-3764-library-home-filter-by-content-type' into jill/FAL-3758-sort-components

yusuf-musleh commented on src/library-authoring/LibraryAuthoringPage.tsx:
```diff
@@ -77,6 +79,7 @@ const LibraryAuthoringPage = () => {
   const handleTabChange = (key: string) => {
     setTabKey(key);
     navigate(key);
+    setSearchParams(searchParams);
   };
```
```suggestion
  const handleTabChange = (key: string) => {
    setTabKey(key);
    navigate({
      pathname: key,
      search: searchParams.toString(),
    });
  };
```

yusuf-musleh replied:
@pomegranited I clicked post accidentally before writing my comment 😅 . The tests were failing seemingly because the search params being updated later was causing issues with the location state for some reason. Putting them inside the `navigate` as one operation fixes the issue, and maintains the functionality we want (thanks @rpenido for the help!).

pomegranited replied:
Thank you so much @yusuf-musleh and @rpenido ! That test failure kept me awake last night, I couldn't figure it out!

pomegranited committed:
fix: navigate to url + params

fixes failing test

pomegranited committed:
test: adds tests for the SearchSortWidget

on the LibraryAuthoringPage.

rpenido commented on src/search-manager/SearchManager.ts:
```diff
@@ -45,6 +91,18 @@ export const SearchContextProvider: React.FC<{
+  const [searchSortOrder, setSearchSortOrder] = useStateWithUrlSearchParam<SearchSortOption>(
+    SearchSortOption.RELEVANCE,
+    'sort',
+    (value: string) => Object.values(SearchSortOption).find((enumValue) => value === `${enumValue}`),
```
nit:
```suggestion
    (value: string) => Object.values(SearchSortOption).find((enumValue) => value === enumValue),
```

rpenido commented on src/search-manager/SearchManager.ts:
```diff
@@ -36,6 +39,49 @@ export interface SearchContextData {
+ * search parameter when returning/setting the state variable.
+ *
+ */
+function useStateWithUrlSearchParam<Type>(
```
Liked this! Good work @pomegranited !

rpenido reviewed (approved):
LGTM :+1: 
Thank you for your great work here @pomegranited. I liked the `useStateWithUrlSearchParam` hook. Very elegant!

- [x] I tested this using the instructions from https://github.com/openedx/frontend-app-course-authoring/pull/1147
- [x] I read through the code
- [x] I checked for accessibility issues
- [x] Includes documentation

pomegranited committed:
fix: stringify enumValue to compare with string

Co-authored-by: Rômulo Penido <romulo@opencraft.com>

yusuf-musleh committed:
feat: Add "Recently Modified" library section (TEMP) (#52)

* feat: Add Recently Modified library section
* feat: Add "View All" to library sections
  The "View All" action appears on sections that pass in a view all action and contain content that exceeds the defined preview limit, which defaults to 4.
* feat: Use intl library section titles
* test: Update tests

pomegranited committed:
fix: style the sort dropdown more like the search filter widgets

There's still a difference with the arrow icon used -- DropDown uses a
css-generated hollow caret, but the filters use the svg solid caret Icon.

pomegranited committed:
feat: Recently Published sort option also excludes unpublished components

pomegranited committed:
fix: removes "Most Relevant" default sort option, use "Clear Filters" instead

pomegranited commented:
Closed in favour of https://github.com/openedx/frontend-app-course-authoring/pull/1147
